arch/sim/src/sim/up_ucontext.c

```diff
@@ -0,0 +1,363 @@
+ *   entry_point  - the entry point for the task
+ *   argc         - the number of cmd arguments
+ *   argv         - array of arguments for the new task
+ *
+ ****************************************************************************/
+
+void up_setup_args_context(void *current_ucontext, void (*entry_point)(void),
+    int argc, char *argv[])
+{
+  makecontext(current_ucontext, entry_point, 2, argc, argv);
+}
+
+/****************************************************************************
+ * Name: up_swap_context
+ *
+ * Description:
+ *   Save the current context in the old_ucontext and activate the
+ *   context from activate_ucontext.
+ *
+ * Input Parameters:
+ *   old_ucontext       - place where we store the current context
+ *   activate_ucontext  - context that we will activate after function
+ *     invocation.
+ *
+ ****************************************************************************/
+
+void up_swap_context(void *old_ucontext, void *activate_ucontext)
+{
+  swapcontext(old_ucontext, activate_ucontext);
+}
+
+/****************************************************************************
+ * Name: up_set_context
+ *
+ * Description:
+ *   Set the current context to the specified ucontext
+ *
+ * Input Parameters:
+ *   current_context - the current context
+ *
+ ****************************************************************************/
+
+void up_set_context(void *current_context)
+{
+  setcontext(current_context);
+}
+
+/****************************************************************************
+ * Name: up_cpu_simulated_interrupt
+ *
+ * Description:
+ *   This function verifies if we are running on the signal handler stack.
+ *
+ * Returned Value:
+ *   Return zero if we are not running on the signal stack otherwise return
+ *   a non zero value.
+ *
+ ****************************************************************************/
+
+int up_cpu_simulated_interrupt(void)
+{
+  uint8_t test_address;
+  unsigned long verify_stack_arg = (unsigned long)&test_address;
+  unsigned long stack_start = (unsigned long)g_ss.ss_sp;
+  unsigned long stack_end;
+
+  if (up_is_stack_downwards(&verify_stack_arg) == true)
+    {
+      stack_end = stack_start - g_ss.ss_size;
+      return stack_start > verify_stack_arg && stack_end < verify_stack_arg;
+    }
+  else
+    {
+      stack_end = stack_start + g_ss.ss_size;
+      return stack_start < verify_stack_arg && stack_end < verify_stack_arg;
+    }
+}
+
+/****************************************************************************
+ * Name: up_setup_sigstackt
+ *
+ * Description:
+ *   This function sets up an alternative stack for the signals from the
+ *   host heap memory.
+ *
+ ****************************************************************************/
+
+void up_setup_sigstack(void)
+{
+  g_ss.ss_sp = malloc(SIGSTKSZ);
+  ASSERT(g_ss.ss_sp != NULL);
+
+  g_ss.ss_size  = SIGSTKSZ;
+  g_ss.ss_flags = 0;
+
+  int ret = sigaltstack(&g_ss, NULL);
```

Comment:
The thread that installs the signal handler with a call to `sigaction` should specify the `SS_ONSTACK` flag.

`
           sa.sa_flags = SA_ONSTACK;
           sa.sa_handler = handler();      /* Address of a signal handler */
           sigemptyset(&sa.sa_mask);
           if (sigaction(SIGSEGV, &sa, NULL) == -1) {
               perror("sigaction");
               exit(EXIT_FAILURE);
           }

`

Reply:
What I mean is that it's fine for non SMP case to just define one interrupt stack for signal handling, but the stack will corrupt in SMP case because the multiple signal handle could be activate at the same time and then utilize the same interrupted stack concurrently.

Reply:
That's right but the behavior is that only the thread that installs the handler with `SS_ONSTACK` will run it on the alternate stack and the others will run the handler on their own stack. Anyways, I think we don't need this alternate stack method to decide if we are running in an interrupt or not. [Your method](https://github.com/xiaoxiang781216/incubator-nuttx/commit/38fcaa6a9cb4893c090d9830a305d4bb209b4b35) seems much better with keeping an `g_cpu_signal` array and marking true/false when we enter a handler.